### PR TITLE
[WIP] Bring back original price in product variants

### PR DESCRIPTION
**Pull request by gorkalaucirica**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets | #7079  |
| License         | MIT |

WIP.

- Admin scenarios are completed (I guess)
- I had an issue with form labels, cannot translate them. Symfony does not support passing variables, any workaround?
- I will really appreciate a code review with some feedback before I add the required features to ShopBundle

Thanks in advance!!

### CODE REVIEW

**Comment by gorkalaucirica** [comment]
I broke some tests... Will try to fix them

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php [hunk 1]
```diff
@@ -492,6 +501,14 @@ public function iChangeItsPriceTo($price, $channelName)
     }
 
     /**
+     * @When /^I change its original price to (?:€|£|\$)([^"]+) for "([^"]+)" channel$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: We usually put prices inside quotation marks, so `(?:€|£|\$)([^"]+)` -> `"(?:€|£|\$)([^"]+)"`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php [hunk 2]
```diff
@@ -1004,6 +1021,20 @@ public function itShouldBePricedAtForChannel(ProductInterface $product, $price,
     }
 
     /**
+     * @Then /^(its|this products) original price should be (?:€|£|\$)([^"]+) for channel "([^"]+)"$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: We usually put prices inside quotation marks, so `(?:€|£|\$)([^"]+)` -> `"(?:€|£|\$)([^"]+)"`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php [hunk 3]
```diff
@@ -1004,6 +1021,20 @@ public function itShouldBePricedAtForChannel(ProductInterface $product, $price,
...
+     * @Then /^(product "[^"]+") original price should be (?:€|£|\$)([^"]+) for channel "([^"]+)"$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: We usually put prices inside quotation marks, so `(?:€|£|\$)([^"]+)` -> `"(?:€|£|\$)([^"]+)"`

File: src/Sylius/Behat/Page/Admin/ProductVariant/CreatePageInterface.php [hunk 4]
```diff
@@ -27,13 +28,19 @@
     public function specifyPrice($price, $channel);
 
     /**
+     * @param int $originalPrice
+     * @param string $channel
+     */
+    public function specifyOriginalPrice($originalPrice, $channel);
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/ProductVariant/CreatePageInterface.php]: `$channel` -> `$channelName`?

File: src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php [hunk 5]
```diff
@@ -68,6 +69,14 @@ public function specifyPrice($channelName, $price)
     /**
      * {@inheritdoc}
      */
+    public function specifyOriginalPrice($channelName, $originalPrice)
+    {
+        $this->getElement('original_price', ['%channel%' => $channelName])->setValue($originalPrice);
```
Comment (Arminek) [inline src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php]: `%channelName%` WDYT?

File: src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/Tab/_details.html.twig [hunk 6]
```diff
@@ -10,7 +10,12 @@
             <div class="two fields">
                 {{ form_row(form.shippingCategory) }}
             </div>
-            {{ form_row(form.channelPricings) }}
+            {% for pricing_form in form.channelPricings %}
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/Tab/_details.html.twig]: `pricing_form` -> `pricingForm`?

File: src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php [hunk 7]
```diff
@@ -49,4 +50,41 @@ function its_price_is_mutable()
...
+        $this->getOriginalPrice()->shouldReturn(null);
+    }
+
+    function its_original_price_should_be_mutable()
+    {
+        $this->setOriginalPrice(2000);
+        $this->getOriginalPrice()->shouldReturn(2000);
+    }
+
+    function its_original_price_should_not_be_lower_than_price()
```
Comment (tuka217) [inline src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php]: Shouldn't you check somehow that that the original price is not lower than the price? Or maybe this function is outdated?

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 8]
```diff
@@ -170,6 +171,14 @@ public function iSetItsPriceTo($price = null, $channel = null)
     }
 
     /**
+     * @When /^I set its original price to "(?:€|£|\$)([^"]+)" for "([^"]+)" channel$/
+     */
+    public function iSetItsOriginalPriceTo($originalPrice = null, $channel = null)
+    {
+        $this->createPage->specifyOriginalPrice($originalPrice, (null === $channel) ? $this->sharedStorage->get('channel') :$channel);
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: 1. We should definitely pass a name of channel two that function, not an object (it would require ``__toString`` method)
2. With this step definition it's not even possible to have ``$originalPrice``` and ``$channel`` as null (as it won't be matched ;))

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php (hunk 3 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: Is this step used anywhere? We should not define steps in advance.

File: src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php [hunk 9]
```diff
@@ -241,7 +250,8 @@ protected function getDefinedElements()
             'images' => '#sylius_product_images',
             'language_tab' => '[data-locale="%locale%"] .title',
             'name' => '#sylius_product_translations_%locale%_name',
-            'price' => '#sylius_product_variant_channelPricings [data-form-collection="item"]:contains("%channel%") input',
+            'price' => '#sylius_product_variant_channelPricings [data-form-collection="item"]:contains("%channel%") input[name$="[price]"]',
+            'original_price' => '#sylius_product_variant_channelPricings [data-form-collection="item"]:contains("%channel%") input[name$="[originalPrice]"]',
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php]: I would keep alphabetical order here ;)

File: src/Sylius/Behat/Page/Admin/Product/UpdateSimpleProductPage.php [hunk 10]
```diff
@@ -427,7 +447,8 @@ protected function getDefinedElements()
             'images' => '#sylius_product_images',
             'language_tab' => '[data-locale="%locale%"] .title',
             'name' => '#sylius_product_translations_%locale%_name',
-            'price' => '#sylius_product_variant_channelPricings [data-form-collection="item"]:contains("%channel%") input',
+            'price' => '#sylius_product_variant_channelPricings [data-form-collection="item"]:contains("%channel%") input[name$="[price]"]',
+            'original_price' => '#sylius_product_variant_channelPricings [data-form-collection="item"]:contains("%channel%") input[name$="[originalPrice]"]',
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/Product/UpdateSimpleProductPage.php]: And here.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php [hunk 11]
```diff
@@ -35,7 +36,12 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 $form = $event->getForm();
 
                 $form->add('price', MoneyType::class, [
-                    'label' => $channel->getName(),
+                    'label' => 'Price for ' . $channel->getName(),
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php]: 1. What is the reason of this change? I know the UI is not perfect for now, but is it really required? (provide some screenshots if you don't mind 😄 )
2. If we need this, there should be a translation key.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php (hunk 8 above)
Comment (Arminek) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: Maybe try to use transformer. WDYT? Right now `$channel` has mixed type.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php (hunk 11 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php]: We should look for some way to allow translating these labels.

File: src/Sylius/Component/Core/Model/ChannelPricingInterface.php [hunk 12]
```diff
@@ -1,12 +1,22 @@
 <?php
 
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
```
Comment (Zales0123) [inline src/Sylius/Component/Core/Model/ChannelPricingInterface.php]: 👍

File: src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php [hunk 13]
```diff
@@ -49,4 +50,41 @@ function its_price_is_mutable()
         $this->setPrice(1000);
         $this->getPrice()->shouldReturn(1000);
     }
+
+    function it_does_not_have_original_price_by_default()
+    {
+        $this->getOriginalPrice()->shouldReturn(null);
+    }
+
+    function its_original_price_should_be_mutable()
```
Comment (Zales0123) [inline src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php]: Avoid word ``should`` in spec names -> ``its_original_price_is_mutable`` would be sufficient

File: src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php [hunk 14]
```diff
@@ -49,4 +50,41 @@ function its_price_is_mutable()
...
+        $this->getOriginalPrice()->shouldReturn(2000);
+    }
+
+    function its_original_price_should_not_be_lower_than_price()
+    {
+        $this->setPrice(2000);
+        $this->setOriginalPrice(1000);
+    }
+
+    function its_price_is_reduced()
```
Comment (Zales0123) [inline src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php]: What about ``its_price_can_be_reduced``? It's not reduced all the time 😄

File: src/Sylius/Behat/Page/Admin/ProductVariant/CreatePage.php [hunk 15]
```diff
@@ -29,7 +30,15 @@ class CreatePage extends BaseCreatePage implements CreatePageInterface
      */
     public function specifyPrice($price, $channel)
     {
-        $this->getDocument()->fillField($channel, $price);
+        $this->getDocument()->fillField('Price for ' . $channel, $price);
```
Comment (Arminek) [inline src/Sylius/Behat/Page/Admin/ProductVariant/CreatePage.php]: `specifyPrice($price, ChannelInterface $channel)`. Instead of concatenation try to use `sprintf('Price for %s', $channel->getName())`.

File: src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php (hunk 7 above)
Comment (Arminek) [inline src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php]: It seams that `ChannelPricing` does not have this behaviour.

File: src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php (hunk 14 above)
Comment (Arminek) [inline src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php]: `its_checks_if_price_has_been_reduced` WDYT?

File: src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php (hunk 5 above)
Comment (gorkalaucirica) [inline src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php]: It's copied from above, can change both

File: src/Sylius/Behat/Page/Admin/ProductVariant/CreatePageInterface.php (hunk 4 above)
Comment (gorkalaucirica) [inline src/Sylius/Behat/Page/Admin/ProductVariant/CreatePageInterface.php]: It's copied from above, can change both

File: src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php (hunk 11 above)
Comment (gorkalaucirica) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php]: As i commented in the PR details I don't know how to add a translation to a label that supports a variable.
Translation string could be Price for %channel% in english but cannot pass to the translator that variable

File: src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php (hunk 14 above)
Comment (gorkalaucirica) [inline src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php]: Whatever you want

File: src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php (hunk 7 above)
Comment (gorkalaucirica) [inline src/Sylius/Component/Core/spec/Model/ChannelPricingSpec.php]: Should we have it? I copied as it was from the PR where the feature was removed, and definitely the test is not correct

**Comment by gorkalaucirica** [comment]
I need to know how to approach to this two things:

- Original price validation, where should be done and constraints to be applied
- Form rendering, any idea to improve it? Injecting the translator to translate the label its a solution that works but maybe is not the best one

File: app/migrations/Version20161214101054.php [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE sylius_channel_pricing ADD originalPrice INT');
```
Comment (inssein) [inline app/migrations/Version20161214101054.php]: I think the convention is for the field to be be named `original_price`

File: src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php (hunk 11 above)
Comment (gabiudrescu) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php]: Maybe this can help: 
http://stackoverflow.com/questions/19593851/symfony2-form-use-translator-for-attr-attribute

Although, it seems this feature is not provided out of the box by symfony: https://github.com/symfony/symfony/pull/10125#issuecomment-112468612

**Comment by gabiudrescu** [comment]
I see originalPrice as following:

 - Default NOT null
 - if it is non existent, then it should be equal to sellingPrice
 - always equal or higher than selling price

This should be implemented as low level as possible. Schema, application level validation, you name it.

I believe this mostly because of separation of comcerns. By having prices as described above, the display layer should always check if those two are equal, then show both or not.

Any other new conditions breaks the separation of concern principle and it rings an alarm bell in my head.

If original price would have been lower than selling price, what would I, as a customer, understand from this? How would that motivate me to buy that product?

Once again, I am telling this because my perception about originalPrice is that we have it because this is the price that gets written on the page with a strikethrough font decoration (and a badge with blinking red saying discount, buy now)

**Comment by pjedrzejewski** [comment]
I agree with @gabiudrescu, original price should be only for displaying purposes. I did not review the code yet, but we definitely want to keep existing logic and only add this as a simple feature.

**Comment by gorkalaucirica** [comment]
@pjedrzejewski So its enough if we leave original_price without validation and we use `isPriceReduced()` to check if a promotion is available in ShopBundle and painting the original page with a strikethrough according to that?

**Comment by crevillo** [comment]
i don't understand the funwction of the original_price combined with the price then. I mean, if we're going to strikethrough the original price and print the price depending on promotions why we need the price itself? price or original price should be enough. 

I would undertand better the original price thing if "price" will be threated as a discount, just like magento does in some ways. something like ~$20.00~ $10.00 from x to y (or something like that)

Actually i feel is not very easy to add different discounts to a lot products, btw...

**Comment by pjedrzejewski** [comment]
@gorkalaucirica Exactly. Catalog wide discounts is a separate feature that we will need to tackle as a separate plugin or maybe it would be possible to implement them without BC breaks. Otherwise things are getting a bit too complicated, while not really solving the problem of catalog wide promotions. (you would still need to edit these manually)

**Comment by gorkalaucirica** [comment]
@pjedrzejewski So we can leave the model and the admin as it is and start working on the shop?

File: src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php (hunk 11 above)
Comment (gorkalaucirica) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php]: Yes, and the don`t plan to support this, so injecting the translator seems the only way to do it

Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Product/ChannelPricingType.php]: As for translation issues:

We can set label for `price` field to "sylius.ui.price", label for `originalPrice` field to "sylius.ui.original_price" and then pass `$channel` to form view to allow for custom rendering. Then we can overwrite Twig block for channel pricing (after #7664 is merged) to something like:

```html
<fieldset>
  <legend>{{ channel.name }}</legend>
  {{ form_row(channelPricing.price) }}
  {{ form_row(channelPricing.originalPrice) }}
</fieldset>
```

If you can't find time to dig into it further, tell me so I'll make a PR to your branch.

File: app/migrations/Version20161214101054.php (hunk 16 above)
Comment (pamil) [inline app/migrations/Version20161214101054.php]: 👍

**Comment by gorkalaucirica** [comment]
@pamil I will rebase and try to apply your suggested changes. Hope this gets into beta2. We've been using it for a couple of months and works great. I think it's a must have

**Comment by gorkalaucirica** [comment]
Rebased and changes made. 

Still to do:

* Form string translation. Maybe @pamil can help here and somehow blocked by #7664 
* Implement this feature in ShopBundle (can be done in another PR)

Anything else?

**Comment by gorkalaucirica** [comment]
Rebased after #7664

**Comment by pamil** [comment]
Form translations made in https://github.com/gorkalaucirica/Sylius/pull/1 :)

**Comment by gorkalaucirica** [comment]
@pamil merged

**Comment by pamil** [comment]
https://github.com/gorkalaucirica/Sylius/pull/2 should make tests passing once again 🎉

**Comment by gorkalaucirica** [comment]
@pamil Merged 😉

**Comment by pjedrzejewski** [comment]
Thanks a lot Gorka! And everyone for review & help. @gorkalaucirica Would you find a moment to have a look at the shop part of the implementation?

**Comment by gorkalaucirica** [comment]
@pjedrzejewski I will give a try this weekend if I find some time.
